pyro/infer/traceenum_elbo.py

```diff
@@ -10,25 +10,65 @@
 import pyro
 import pyro.ops.jit
 import pyro.poutine as poutine
-from pyro.distributions.util import is_identically_zero
+from pyro.distributions.util import broadcast_shape, is_identically_zero
 from pyro.infer.elbo import ELBO
 from pyro.infer.enum import get_importance_trace, iter_discrete_escape, iter_discrete_extend
 from pyro.infer.util import Dice, is_validation_enabled
+from pyro.ops.einsum import shared_intermediates
+from pyro.ops.sumproduct import sumproduct
+from pyro.poutine.enumerate_messenger import EnumerateMessenger
 from pyro.util import check_traceenum_requirements, warn_if_nan
 
 
+# TODO move this logic into a poutine
+def _compute_model_costs(model_trace, guide_trace, ordering):
+    # Collect model log_probs, possibly marginalizing out enumerated model variables.
+    costs = OrderedDict()
+    enum_logprobs = OrderedDict()
+    enum_dims = []
+    for name, site in model_trace.nodes.items():
+        if site["type"] == "sample":
+            site["log_prob"]._pyro_name = name  # DEBUG
+            if name in guide_trace or not site["infer"].get("_enumerate_dim") is not None:
+                costs.setdefault(ordering[name], []).append(site["log_prob"])
+            else:
+                enum_logprobs.setdefault(ordering[name], []).append(site["log_prob"])
+                enum_dims.append(len(site["fn"].event_shape) - len(site["value"].shape))
+    if not enum_logprobs:
+        return costs
+
+    # Marginalize out all enumerated variables.
+    enum_boundary = max(enum_dims) + 1
+    marginal_costs = OrderedDict((t, []) for t in costs)
+    with shared_intermediates():
+        for t, costs_t in costs.items():
+            # TODO refine this coarse dependency ordering using time and tensor shapes.
+            logprobs_t = sum((logprobs_u for u, logprobs_u in enum_logprobs.items() if u <= t), [])
+            if not logprobs_t:
+                marginal_costs[t] = costs_t
+                continue
+
+            ordinal_shape = broadcast_shape(*(x.shape for x in logprobs_t))
+            for cost_t in costs_t:
+                target_shape = broadcast_shape(cost_t.shape, ordinal_shape)
+                target_shape = target_shape[enum_boundary:] if enum_boundary else ()
+                print('{}: target_shape={}, factor shapes = {}'.format(
```

Comment:
Can we remove this print statement, or add it as `debug` statement using `pyro.log.debug` instead? Our test logs are increasing in size, and it will be nice to be able to toggle debug statements if needed.